test/MTLSpec.hs

```diff
@@ -0,0 +1,143 @@
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE TypeApplications    #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+module MTLSpec where
+
+import           Polysemy
+import           Polysemy.Reader
+import           Polysemy.Writer
+import           Polysemy.State
+import           Polysemy.Error
+import           Polysemy.MTL
+
+import qualified Data.Text                     as T
+import           Test.Hspec
+
+import           Control.Monad                  ( replicateM_
+                                                , when
+                                                )
+import qualified Control.Monad.Reader.Class    as S
+import qualified Control.Monad.Writer.Class    as S
+import qualified Control.Monad.State.Class     as S
+import qualified Control.Monad.Error.Class     as S
+
+
+{-
+We could re-write these to use polysemy directly.  Imagine, though
+that these come from external libraries so you can't so easily
+re-write them.
+-}
+getEnvLength :: S.MonadReader T.Text m => m Int
+getEnvLength = S.ask >>= return . T.length
+
+replicateTell :: S.MonadWriter [Int] m => Int -> Int -> m ()
+replicateTell n m = replicateM_ n $ S.tell [m]
+
+retrieveAndUpdateN :: S.MonadState Int m => Int -> m Int
+retrieveAndUpdateN n = do
+  m <- S.get
+  S.put n
+  return m
+
+-- this one is exceptionally boring
+throwOnZero :: S.MonadError T.Text m => Int -> m Int
+throwOnZero n = do
+  when (n == 0) $ S.throwError "Zero!"
+  return n
+
+someOfAll
+  :: (S.MonadReader T.Text m, S.MonadWriter [Int] m, S.MonadState Int m)
+  => m T.Text
+someOfAll = do
+  n <- S.get
+  S.tell [n]
+  S.ask
+
+spec :: Spec
+spec = do
+  describe "MTL" $ do
+
+    it
+        "should return 10, the sum of lengths of the strings provided to run (\"Text1\") and then to local (\"Text2\")"
+      $ do
+          flip shouldBe 10 $ run . runReader "Text1" $ do
+            a <- absorbReader getEnvLength
+            b <- local (const "Text2") $ absorbReader getEnvLength
+            return (a + b)
+
+    it
+        "should return the sum, after censoring, of all things told.  In this case, 16, the sum of \"init [1,5,5,5,5]\""
+      $ do
+          flip shouldBe 16 $ sum . fst . run . runWriter $ do
```

Comment:
Can we have the same test as this, but with

```haskell
run . runWriter . absorbWriter $ do
```

and then just run the same test in the `MoandWriter` monad?

Reply:
As an additional test? OKay.

Reply:
Ah! This caught a bug.  Thanks!  The differences between the mtl-writer and the polysemy writer are weird.  To fix the bug I had to do an awkward thing when implementing ```pass```: I had to use ```listen``` to get ```w``` so far and then hand censor ```tell w >> pure a``` which seems...odd.